mindmeld/models/embedder_models.py

```diff
@@ -0,0 +1,184 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains the embedder model class.
+"""
+from abc import ABC, abstractmethod
+import pickle
+import logging
+import os
+import string
+import numpy as np
+
+from .. import path
+from .helpers import register_embedder
+from .taggers.embeddings import WordSequenceEmbedding
+
+
+logger = logging.getLogger(__name__)
+
+try:
+    from sentence_transformers import SentenceTransformer
+
+    register_bert = True
+except ImportError:
+    logger.warning("Must install the extra [bert] to use the built in embbedder.")
+    register_bert = False
+
+
+class Embedder(ABC):
+    """
+    Base class for embedder model
+    """
+
+    def __init__(self, app_path, **kwargs):
+        """Initializes an embedder.
+        """
+        self.model_name = kwargs.get("model_name", "default")
+        self.cache_path = path.get_embedder_cache_file_path(
+            app_path, kwargs.get("embedder_type", "default"), self.model_name
+        )
+
+        folder = os.path.dirname(self.cache_path)
+        if not os.path.isdir(folder):
+            os.makedirs(folder)
+
+        if not os.path.exists(self.cache_path) or os.path.getsize(self.cache_path) == 0:
+            self.cache = {}
+        else:
+            with open(self.cache_path, "rb") as fp:
+                self.cache = pickle.load(fp)
+
+        self.model = self.load(**kwargs)
+
+    @abstractmethod
+    def load(self, **kwargs):
+        """Loads the embedder model
+
+        Returns:
+            The model object.
+        """
+        pass
+
+    @abstractmethod
+    def encode(self, text_list):
+        """
+        Args:
+            text_list (list): A list of text strings for which to generate the embeddings.
+
+        Returns:
+            (list): A list of numpy arrays of the embeddings.
+        """
+        pass
+
+    def clear_cache(self):
+        """Deletes the cache file.
+        """
+        if os.path.exists(self.cache_path):
+            os.remove(self.cache_path)
+
+    def get_encodings(self, text_list):
+        """Fetches the encoded values from the cache, or generates them.
+        Args:
+            text_list (list): A list of text strings for which to get the embeddings.
+
+        Returns:
+            (list): A list of numpy arrays with the embeddings.
+        """
+        encoded = [self.cache.get(text, None) for text in text_list]
+        cache_miss_indices = [i for i, vec in enumerate(encoded) if vec is None]
+        text_to_encode = [text_list[i] for i in cache_miss_indices]
+        model_encoded_text = self.encode(text_to_encode)
+
+        for i, v in enumerate(cache_miss_indices):
+            encoded[v] = model_encoded_text[i]
+            self.cache[text_to_encode[i]] = model_encoded_text[i]
+        return encoded
+
+    def dump(self):
```

Comment:
should we pass in an optional path where you can dump your model elsewhere outside of the cache?

Reply:
We don't do this for the dump methods of our other components, so this seems more consistent?

Reply:
I think the other models were pretty small and standard; if the user wants to inspect or analyze the embedding in some ways it might be convenient to allow them to dump to a custom location.

Reply:
This is just a cache of embeddings, not the model itself, so I don't think it would be useful to inspect/analyze?

Reply:
I see, OK